snapcraft_legacy/internal/project_loader/__init__.py

```diff
@@ -53,3 +62,38 @@ def replace_attr(
         return result
 
     return attr
+
+
+def _validate_replacement(attr: str, replacement: str, value: Optional[str]) -> None:
+    """Validate if a replacement can occur for an attribute.
+
+    Some replacement data cannot be used if it is not defined, such as the build-for
+    arch and build-for arch triplet. If the value of these replacements is None, then
+    a validation error is raised.
+
+    :param attr: String that may contain data to replace.
+    :param replacement: Project variable to replace.
+    :param value: New value for replacement.
+
+    :raises Exception: If a replacement cannot occur.
+    """
+    project_variables = [
+        "SNAPCRAFT_ARCH_BUILD_FOR",
+        "SNAPCRAFT_ARCH_TRIPLET_BUILD_FOR",
+    ]
+
+    # expand to shell syntax for variables (`$item` and `${item}`)
+    expanded_project_variables = (
+        [f"${item}" for item in project_variables ] +
+        [f"${{{item}}}" for item in project_variables]
+    )
```

Comment:
```suggestion
    expanded_project_variables = [
        *(f"${item}" for item in project_variables),
        *(f"${{{item}}}" for item in project_variables),
    ]
```

This is more obvious what it's doing IMO, but no strong feelings

Reply:
Implemented, thanks!

Reply:
Why a list and not a tuple in the refactored version?

Reply:
Good point, a tuple is more appropriate.